### PR TITLE
Provider demo nosetest

### PR DESCRIPTION
**Pull request by tzhang-ext-aa**
`nosetests -sv pypact/demo/provider`

still need to fix mock issues.

### CODE REVIEW

File: pypact/contract_factory.py [hunk 1]
```diff
@@ -2,39 +2,84 @@
...
-from pypact.service import MockService
+from pypact.interaction import Interaction
 
-mock_servers = {}
+contracts = {}
 
 PACT_PATH = 'pact'
 
 
+class Contract(object):
```
Comment (chan-aa) [inline pypact/contract_factory.py]: Contract class should be in a dedicated class file

File: pypact/service.py [hunk 2]
```diff
@@ -26,25 +26,6 @@ def __init__(self, scheme='http', domain='localhost', port=80):
     def get_service_host(self):
         return "{}://{}".format(self.scheme,self.netloc)
```
Comment (chan-aa) [inline pypact/service.py]: also need update the UT 
pypact/pypact/test

File: README.md [hunk 3]
```diff
@@ -172,21 +172,8 @@ Script will auto generate contract file at pypact/pact/${provider_name}/${consum
...
-    def setUp(self):
-        # prepare all the data for the contract with MyClient
-        
-    def tearDown(self):
-        pass
 
-    def test_contract(self):
-    	mock_client = MockClient('./contract_with_myclient')  # load contract file
-        mock_client.test(app)   
+#### run provider cases
```
Comment (chan-aa) [inline README.md]: More detail required to illustrate how to write provider contract test

File: pypact/demo/provider/provider_test_base.py [hunk 4]
```diff
@@ -43,26 +46,35 @@ class ProviderTestBase(object):
     and has the "test_scope" within the class definition.
     """
     @abstractproperty
-    def test_scope(self):
+    def providers(self):
```
Comment (luyun-aa) [inline pypact/demo/provider/provider_test_base.py]: I think this property is not really needed. For a service, the provider name is known and fixed.

File: pypact/demo/provider/provider_test_base.py [hunk 5]
```diff
@@ -43,26 +46,35 @@ class ProviderTestBase(object):
...
+            # import ipdb; ipdb.set_trace()
+            if isinstance(self.include_states, basestring):
+                self.include_states = [self.include_states]
+            for include_path in self.include_states:
+                importlib.import_module(include_path)
+
+        if isinstance(self.providers, tuple):
+            self.providers = [self.providers]
+        for provider_name, test_client in self.providers:
+            for consumer_name in contract_factory.contracts[provider_name]:
```
Comment (luyun-aa) [inline pypact/demo/provider/provider_test_base.py]: This `for` loop means that every provider test will test all contracts? This doesn't make sense. A contract test case on the provider side should test one contract with one consumer.

File: pypact/demo/provider/provider_test_base.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    service developer shall write their own Test class that derived from this class,
+    and has the "test_scope" within the class definition.
+    """
+    @abstractproperty
+    def providers(self):
+        return [
+            ("CategoryService_Provider", None) # the test client
+        ]
+
+    def test_case_generator(self):
```
Comment (luyun-aa) [inline pypact/demo/provider/provider_test_base.py]: This naming is confusing, please rename this method to something like `test_contract`.

File: pypact/demo/provider/provider_test_base.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class ProviderTestBase(object):
+    """
+    the base class that perform the testing.
+    service developer shall write their own Test class that derived from this class,
+    and has the "test_scope" within the class definition.
+    """
+    @abstractproperty
+    def providers(self):
```
Comment (luyun-aa) [inline pypact/demo/provider/provider_test_base.py]: I think this property is not really needed. For a service, the provider name is known and fixed.

File: README.md [hunk 8]
```diff
@@ -169,24 +167,67 @@ Script will auto generate contract file at pypact/pact/${provider_name}/${consum
...
 }
 ```
+
 ## Provider Usage
 
-### Write a test for microservice
+Currently we only support services(providers) that are built with `flask`. And the provider test is built upon `flask.test_client`.
+
+### prapear provider state
+Different `interactiion` requires different `provider state`, as a result we have to have a mechanism to setup/teardown the states.
```
Comment (chan-aa) [inline README.md]: interaction

File: README.md [hunk 9]
```diff
@@ -169,24 +167,67 @@ Script will auto generate contract file at pypact/pact/${provider_name}/${consum
...
+                33,
+                444
+            ]
+        },
+    })
+```
+
+### prepare the case entry
+```python
+from pypact.demo.provider.provider_test_base import ProviderTestBase
```
Comment (chan-aa) [inline README.md]: How can we detect there are some case missing in the contract?

Comment (chan-aa) [inline README.md]: I still prefer provider should run in this way.
1. prepare env
2. read contract file
3. detect case pass or failed(failed case is response are unexpected, include give wrong response, and not provide test necessary data in prepare steps )

File: README.md [hunk 10]
```diff
@@ -15,17 +15,15 @@ For now, the only supported protocol is HTTP, but the principles of pact applies
 ## Example
 ![Diagram](https://github.com/luyun-aa/pypact/raw/master/MyClientAndRankService.png)
 
-
-
 ## Consumer Usage (MyClient)
 [Contract Test Consumer Demo](https://github.com/luyun-aa/pypact/tree/make_contract_file/pypact/demo/consumer)
+
 ### The client service
```
Comment (chan-aa) [inline README.md]: After finalize all design. we'd better provide another illustration how it works for multi consumers and multi providers

File: pypact/demo/consumer/test_2_provider_consumer/test_topchart_category.py [hunk 11]
```diff
@@ -27,10 +30,10 @@ def test_topchart_service(self):
 
 class RankServiceTest(TestCase):
     def test_rank_service(self):
-        service = contract_factory.register("TopChartService_Provider", "TopChartServiceTest_Consumers")
-        (service
+        contract = contract_factory.register("TopChartService_Provider", "TopChartCategoryServiceTest_Consumer")
+        (contract
             .given("Top chart on 2017-03-17")
```
Comment (chan-aa) [inline pypact/demo/consumer/test_2_provider_consumer/test_topchart_category.py]: should it be a different given status?

File: pypact/demo/consumer/test_2_provider_consumer/test_topchart_category.py [hunk 12]
```diff
@@ -40,7 +43,27 @@ def test_rank_service(self):
...
+        service = MockService(contract)
         with service:
             TopChartService.base_uri = service.get_service_host()
             ranks = topchart_service.get_ranks()
             self.assertEqual([1, 22, 33, 444], ranks)
+
+
+class CategoryServiceTest(TestCase):
+    def test_category_service(self):
+        contract = contract_factory.register("CategoryService_Provider", "TopChartCategoryServiceTest_Consumer")
```
Comment (chan-aa) [inline pypact/demo/consumer/test_2_provider_consumer/test_topchart_category.py]: Dose this case aim to illustrate different consumer share with same contract test?

File: README.md [hunk 13]
```diff
@@ -165,28 +163,71 @@ Script will auto generate contract file at pypact/pact/${provider_name}/${consum
...
+
 ```python
-from unittest import TestCase
-from pypact import MockClient
-from application import app  # flask app
+from pypact.demo.provider.db_manager import MockedDBManager, mocked_db_managers
+from pypact.demo.provider.provider_test_base import provider_state
+
+
+@provider_state("Top chart on 2017-03-17")
```
Comment (luyun-aa) [inline README.md]: Inconsistent with real code

File: README.md [hunk 14]
```diff
@@ -169,24 +167,67 @@ Script will auto generate contract file at pypact/pact/${provider_name}/${consum
...
-    	mock_client = MockClient('./contract_with_myclient')  # load contract file
-        mock_client.test(app)   
+### run provider cases
+```bash
+nosetests -sv pypact/demo/provider
 ```
+
+# Try it!
+```bash
+nosetests -sv pypact/demo/consumer && nosetests -sv pypact/demo/provider
```
Comment (chan-aa) [inline README.md]: for provider the only input is the contract.
and different provider the input and test are dedicated.
it is not a good example.

File: pypact/demo/provider/product_service/provider_states.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+                "balabala": "hahahaha",
+            }
+        }
+    })
+    httpretty.enable()
+    httpretty.HTTPretty.allow_net_connect = False
+    httpretty.register_uri(httpretty.GET, "http://product_service/product", body="")
+
+
+@product_wechat_on_2015_12_12.teardown
```
Comment (luyun-aa) [inline pypact/demo/provider/product_service/provider_states.py]: Could we do this in a more elegant way?

**Comment by luyun-aa** [comment]
I'm merging this PR. Please move to aa-pypact repo to continue.

File: README.md [hunk 16]
```diff
@@ -46,26 +44,26 @@ class TopChartService(BaseService):
...
 
 class TopChartServiceTest(TestCase):
     def test_topchart_service(self):
-        service = contract_factory.register("TopChartService_Provider", "TopChartServiceTest_Consumers")
-        (service
+        contract = contract_factory.register("TopChartServiceProvider", "TopChartCategoryServiceTestConsumer")
+        (contract
             .given("Top chart on 2017-03-17")
             .upon_receiving("a request for top chart")
             .with_request(method="get", path="/topchart")
```
Comment (luyun-aa) [inline README.md]: give a example with arguments in query string

File: pypact/demo/provider/product_service/test_product_provider.py [hunk 17]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2017 App Annie Inc. All rights reserved.
+# -*- coding: utf-8 -*-
+from pypact.demo.provider.product_service import product_service
+from pypact.demo.provider.provider_test_base import ProviderTestBase
+
+
+class TestProductProvider(ProviderTestBase):
+    provider = "ProductServiceProvider", product_service.app.test_client()
+    include_states = "pypact.demo.provider.product_service.provider_states"
```
Comment (luyun-aa) [inline pypact/demo/provider/product_service/test_product_provider.py]:
```
scripts/provider_test.py -A application.app --provider-name ProductServiceProvider --provider-states path/to/states/dir --consumer RankService
```

**Pull request merged**